libraries/AP_EFI/AP_EFI_Serial_Hirth.cpp

```diff
@@ -0,0 +1,328 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <AP_HAL/AP_HAL.h>
+
+#include "AP_EFI_Serial_Hirth.h"
+
+#if HAL_EFI_ENABLED
+#include <AP_SerialManager/AP_SerialManager.h>
+#include "SRV_Channel/SRV_Channel.h"
+
+
+/**
+ * @brief Constructor with port initialization
+ * 
+ * @param _frontend 
+ */
+AP_EFI_Serial_Hirth::AP_EFI_Serial_Hirth(AP_EFI &_frontend) : AP_EFI_Backend(_frontend) {
+    port = AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_EFI, 0);
+}
+
+
+/**
+ * @brief checks for response from or makes requests to Hirth ECU periodically
+ * 
+ */
+void AP_EFI_Serial_Hirth::update() {
+    bool status = false;
+    uint32_t now = AP_HAL::millis();
+
+    //log the delay in loop interval
+    internal_state.loop_cnt = now - last_loop_ms;
+    last_loop_ms = now;
+
+    if ((port != nullptr)) {
+
+        //waiting for response
+        if (waiting_response) {
+
+            // reset request if not response for SERIAL_WAIT_TIMEOUT-ms
+            if (now - last_response_ms > SERIAL_WAIT_TIMEOUT) {
+                waiting_response = false;
+                port->discard_input();
+                last_response_ms = now;
+                internal_state.ack_fail_cnt++;
+                if(data_send == 1) {internal_state.ack_s1++;}
+                if(data_send == 2) {internal_state.ack_s2++;}
+                if(data_send == 3) {internal_state.ack_s3++;}
+                if(data_send == 5) {internal_state.ack_thr++;}
+            }
+            else{
+
+                num_bytes = port->available();
+
+                // if already requested
+                if (num_bytes >= expected_bytes) {
+                    
+                    // read data from buffer
+                    uint8_t computed_checksum = 0;
+                    computed_checksum += res_data.quantity = port->read();
+                    computed_checksum += res_data.code = port->read();
+
+                    if (res_data.code == req_data.code) {
+                        for (int i = 0; i < (res_data.quantity - QUANTITY_REQUEST_STATUS); i++) {
+                            computed_checksum += raw_data[i] = port->read();
+                        }
+                    }
+
+                    res_data.checksum = port->read();
+                    computed_checksum = computed_checksum % 256;
```

Comment:
```suggestion
```

You're not going to have a value above 256 in a `uint8_t`

Reply:
code modified